### PR TITLE
GH-91719: Make MSVC generate somewhat faster switch code

### PR DESCRIPTION
**Pull request by gvanrossum**
Apparently a switch on an 8-bit quantity where all cases are
present generates a more efficient jump (doing only one indexed
memory load instead of two).

See https://github.com/faster-cpython/ideas/issues/321#issuecomment-1103263673

### CODE REVIEW

**Comment by markshannon** [comment]
Would it make more sense to redefine `opcode` to be a `uint8_t`, rather than casting it?

We should probably make `use_tracing` an 8 bit unsigned integer as well.

**Comment by gvanrossum** [comment]
> Would it make more sense to redefine `opcode` to be a `uint8_t`, rather than casting it?

Yeah, I had considered that, it makes sense. I'll confirm that it has the same effect.

> We should probably make `use_tracing` an 8 bit unsigned integer as well.

I don't see why -- it's not used in a similar switch AFAICT, and it's not cramped for space in its struct. I assume for most other operations the cost of loading an int and loading a byte is effectively the same, since the CPU has to load a whole cache line (32 or 64 bytes) anyway.

**Comment by gvanrossum** [comment]
I don't believe this needs a news blurb.

**Comment by markshannon** [comment]
> I don't see why -- it's not used in a similar switch AFAICT, and it's not cramped for space in its struct. I assume for most other operations the cost of loading an int and loading a byte is effectively the same, since the CPU has to load a whole cache line (32 or 64 bytes) anyway.

The dispatch sequence includes `opcode |= cframe.use_tracing`.
If `cframe.use_tracing` is a 32 bit int, then the compiler needs to add a cast.
If it is the same type as `opcode`, it does not.

**Comment by gvanrossum** [comment]
@markshannon, please re-review. I confirmed that the switch still uses a single indirection (`goto *(base + offset_table[opcode])`). I also found that the `opcode |= use_tracing` is a single instruction (but maybe it always was one?).

File: Python/ceval.c [hunk 1]
```diff
@@ -1774,6 +1774,8 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
     {
 #else
     dispatch_opcode:
+        // The cast, combined with the EXTRA_CASES macro,
+        // causes MSVC to generate a faster switch.
```
Comment (erlend-aasland) [inline Python/ceval.c]: Just a heads-up that this comment now needs a rewording (and possibly a relocation)

File: Tools/scripts/generate_opcode_h.py [hunk 2]
```diff
@@ -129,6 +129,13 @@ def main(opcode_py, outfile='Include/opcode.h'):
         fobj.write("};\n")
         fobj.write("#endif\n")
 
+        fobj.write("\n")
+        fobj.write("#define EXTRA_CASES \\\n")
+        for i, flag in enumerate(used):
+            if not flag and i != 255:
```
Comment (erlend-aasland) [inline Tools/scripts/generate_opcode_h.py]: 255 is explicitly marked as used a few lines earlier.

```suggestion
            if not flag:
```

**Comment by gvanrossum** [comment]
I wonder if the new `-_Py_DECREF_SPECIALIZED` also needs to be macrofied?

**Comment by neonene** [comment]
This PR:
```
                                     entry  static
Function Name                        count   instr
_PyEval_EvalFrameDefault          21144421    7914
_Py_atomic_load_64bit_impl       589467249      28
Py_DECREF                        490380800       5
Py_IS_TYPE                       290978861       7
Py_XDECREF                       284071989       5
is_method                        164652084      13
_Py_atomic_load_32bit_impl       162393425      28
_Py_DECREF_SPECIALIZED           102609766       5  <<<
```

**Comment by markshannon** [comment]
Looks good to me

**Comment by vstinner** [comment]
Oh wow, that's a simple and clever optimization! Great that it helps MSVC to optimize Python on Windows!

**Comment by vstinner** [comment]
Follow-up fo clean the public API: https://github.com/python/cpython/pull/91906

**Pull request merged**